### ISSUE TITLE
Specifying output separately to generation script

### ISSUE THREAD

**Issue by squaremo** [jkcfg/jk#125]
While making some examples in github.com/jkcfg/kubernetes, I found myself wanting to specify the (style of) output of a script from outside the script. For instance, the chart example generates a set of resources, which could be output as a single YAML stream to stdout, or written into YAML files.

Since it's just code, of course you can write whatever you like; but when making general tooling, it seems useful to separate the generation of the objects from the output.

Some not necessarily mutually-exclusive ideas @dlespiau and I thought of:

#### a "generator" protocol whereby the generating script returned `Promise [resource]`, which is passed on to something else to output

 - another possibility: Promise `{ filename: resource, ...}`

#### have a different subcommand, with flag(s) for controlling output

e..g,

    jk generate ./chart.js --output-stream

.. possibly using the protocol from above, to convey the generated resources. One trick here is that it doesn't always make sense to say, output as a stream -- also, as a stream of what? What if one of the things is actually JSON, or a string?

#### specify a module to do the output (and the generation?)

    jk run ./index.js --output-module @jkcfg/kubernetes/write-stream

**Comment by dlespiau** [comment]
I'm warming up to the idea of a new `generate` command with a well defined way to pass the list of files, `std.write` options and values. We could accept various forms of that protocol:

- a list of `{ file: 'foo.yaml', content: value, format: std.Format.JSON, .... }`
- a map of `{ filename: resource, ...}`
- a promise returning one of the  two things above

**Comment by bryanlarsen** [comment]
A couple of other "would be nice" considerations:

- consideration to hierarchical designs.  We have a parent-child setup: the parent.js calls childA.js and childB.js.    It would be awesome if we could run `jk generate` on either the parent or on a child and have it do the right thing.    Given that jk is plain old javascript, parent.js will end up looking something like `export default [].concat(childA, childB)` so this should flow fairly naturally

- parameter passing.   Right now all of our children are implemented as functions that take parameters from the parent.   We don't necessarily require a way to pass parameters to the children when we're standalone -- defaults are sufficient for our use case in that situation, but we do require a way to pass parameters from parents to children.    So it would be real nice if the generate protocol was a function that returned a promise to a list or map.

**Comment by dlespiau** [comment]
I've merged some support of this with a new `generate` command:

- A simple example can be found here: https://github.com/jkcfg/jk/tree/master/examples/quick-start/alice
- The default value doesn't have to be an array, it can be a Promise to one: https://github.com/jkcfg/jk/blob/master/tests/generate-promise.js
- Supports outputting files on stdout (`--stdout`) as long as all the files are either YAML or JSON files.

@bryanlarsen I wasn't sure if it's what you had in mind for promises. If not, don't hesitate to file a new issue with what you'd like to be able to achieve.

**Comment by bryanlarsen** [comment]
@dlespiau

Here is an example of how I would use it with your PR:

generate-child.js:
```
export function child(param = 'some default') {
  return [{file: 'foo.yaml', value:{ param }}];
}

export default child();
```

generate-parent.js:
```
import { child } from './generate-child.js';

export default child('from-parent');
```

This let's me do a jk generate on either the parent or the child.   (Imagine the parent having 12 children each outputting a small handful of files)

You'll notice how I'm exporting two different things from generate-child.js.  It'd be nice if I could just export that function as the default.   So when a default export to generate is a function, just evaluate the function and then apply your existing logic vis-a-vis promises.

In other words, it'd be nice if `jk generate` worked on this:

```
export default function child(param = 'some default') {
  return [{file: 'foo.yaml', value:{ param }}];
}
```

But it's no big deal.   The code at the top certainly meets my needs quite nicely.

**Comment by dlespiau** [comment]
Ah yes, that makes sense! Added support for default value being functions in #197.

```console
$ cat child.js 
export default function child(param = 'some default') {
  return [{file: 'foo.yaml', value:{ param }}];
}

$ cat parent.js 
import child from './child.js';

export default child('from-parent');

$ jk generate --stdout child.js 
param: some default
$ jk generate --stdout parent.js 
param: from-parent
```
